pkg/policy/selectorcache.go

```diff
@@ -196,12 +170,182 @@ func (l *labelIdentitySelector) updateSelections() {
 	sort.Slice(selections, func(i, j int) bool {
 		return selections[i] < selections[j]
 	})
-	l.setSelections(&selections)
+	s.setSelections(&selections)
+}
+
+func (s *selectorManager) setSelections(selections *[]identity.NumericIdentity) {
+	if len(*selections) > 0 {
+		atomic.StorePointer(&s.selections, unsafe.Pointer(selections))
+	} else {
+		atomic.StorePointer(&s.selections, unsafe.Pointer(&emptySelection))
+	}
+}
+
+// lock must be held
+func (s *selectorManager) notifyUsers(added, deleted []identity.NumericIdentity) {
+	for user := range s.users {
+		user.IdentitySelectionUpdated(s, s.GetSelections(), added, deleted)
+	}
 }
 
 type fqdnSelector struct {
-	labelIdentitySelector
-	// fqdnSelections map[string]identity.NumericIdentity // identity.String(): "cidr:1.1.1.1" -> identity of 1.1.1.1
+	selectorManager
+	selector api.FQDNSelector
+}
+
+type labelIdentitySelector struct {
+	selectorManager
+	selector api.EndpointSelector
+}
+
+//
+// CachedSelector implementation (== Public API)
+//
+// No locking needed.
+//
+
+//
+// identitySelector implemenentation (== internal API)
+//
+
+// lock must be held
+func (l *labelIdentitySelector) notifyUsers(added, deleted []identity.NumericIdentity) {
+	for user := range l.users {
+		user.IdentitySelectionUpdated(l, l.GetSelections(), added, deleted)
+	}
+}
+
+// UpdateFQDNSelector updates the mapping of fqdnKey (the FQDNSelector from a
+// policy rule as a string) to to the provided list of identities. If the contents
+// of the cachedSelections differ from those in the identities slice, all
+// users are notified.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity) {
+
+	fqdnKey := fqdnSelec.String()
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+
+	var fqdnSel *fqdnSelector
+
+	selector, exists := sc.selectors[fqdnKey]
+	if !exists || selector == nil {
+		fqdnSel = &fqdnSelector{
+			selectorManager: selectorManager{
+				key:              fqdnKey,
+				users:            make(map[CachedSelectionUser]struct{}),
+				cachedSelections: make(map[identity.NumericIdentity]struct{}),
+			},
+			selector: fqdnSelec,
+		}
+		sc.selectors[fqdnKey] = fqdnSel
+	} else {
+		fqdnSel = selector.(*fqdnSelector)
+	}
+
+	// Convert identity slice to map for comparison with cachedSelections map.
+	idsAsMap := make(map[identity.NumericIdentity]struct{}, len(identities))
+	for _, v := range identities {
+		idsAsMap[v] = struct{}{}
+	}
+
+	var added, deleted []identity.NumericIdentity
+
+	/* TODO - the FQDN side should expose what was changed (IPs added, and removed)
+	*  not all IPs corresponding to an FQDN - this will make this diff much
+	*  cheaper, but will require more plumbing on the FQDN side. for now, this
+	*  is good enough.
+	*
+	*  Case 1: identities did correspond to this FQDN, but no longer do. Reset
+	*  the map
+	 */
+	if len(identities) == 0 && len(fqdnSel.cachedSelections) != 0 {
```

Comment:
This case is basically the default case 3, right? since that would iterate over cachedSelection, deleting each one sequentially because idAsMap is empty. The range identities would also be a no-op.
This is more to make sure I understand it. I think the separate case is more clear.

Reply:
Correct!